File: java/Samples/Biometrics/VB/ServerSampleVB/Connections/DirectoryEnumerator.vb
Imports System
Imports System.IO
Imports Neurotec.Biometrics
Imports Neurotec.IO

Public Class DirectoryEnumerator
	Implements ITemplateLoader
#Region "Private fields"

	Private ReadOnly _lockObject As Object = New Object()
	Private ReadOnly _files() As String

	Private _index As Integer = -1

#End Region

#Region "Public constructor"

	Public Sub New(ByVal directory As String)
		If directory Is Nothing OrElse (Not System.IO.Directory.Exists(directory)) Then
			Throw New ArgumentException("Directory doesn't exist")
		End If

		Dir = directory
		_files = System.IO.Directory.GetFiles(Dir)
	End Sub

#End Region

#Region "Public properties"

	Private privateDir As String
	Public Property Dir() As String
		Get
			Return privateDir
		End Get
		Private Set(ByVal value As String)
			privateDir = value
		End Set
	End Property

#End Region

#Region "ITemplateLoader members"

	Public Sub BeginLoad() Implements ITemplateLoader.BeginLoad
		SyncLock _lockObject
			If _index <> -1 Then
				Throw New InvalidOperationException()
			End If
			_index = 0
		End SyncLock
	End Sub

	Public Sub EndLoad() Implements ITemplateLoader.EndLoad
		SyncLock _lockObject
			_index = -1
		End SyncLock
	End Sub

	Public Function LoadNext(<System.Runtime.InteropServices.Out()> ByRef subjects() As NSubject, ByVal n As Integer) As Boolean Implements ITemplateLoader.LoadNext
		SyncLock _lockObject
			If _index = -1 Then
				Throw New InvalidOperationException()
			End If
			subjects = Nothing
			If _files.Length = 0 OrElse _files.Length <= _index Then
				Return False
			End If

			Dim count As Integer = _files.Length - _index
			count = If(count > n, n, count)
			subjects = New NSubject(count - 1) {}
			For i As Integer = 0 To count - 1
				Dim file As String = _files(_index)
				_index += 1
				subjects(i) = New NSubject()
				subjects(i).SetTemplateBuffer(New NBuffer(System.IO.File.ReadAllBytes(file)))
				subjects(i).Id = Path.GetFileNameWithoutExtension(file)
			Next i
			Return True
		End SyncLock
	End Function

	Public Sub Dispose() Implements ITemplateLoader.Dispose
	End Sub

	Public ReadOnly Property TemplateCount() As Integer Implements ITemplateLoader.TemplateCount
		Get
			Return _files.Length
		End Get
	End Property

#End Region

#Region "Public methods"

	Public Overrides Function ToString() As String
		Return Dir
	End Function

#End Region
End Class

File: java/Samples/Biometrics/CS/ServerSampleCS/Connections/DirectoryEnumerator.cs
using System;
using System.IO;
using Neurotec.Biometrics;
using Neurotec.IO;

namespace Neurotec.Samples.Connections
{
	public class DirectoryEnumerator : ITemplateLoader
	{
		#region Private fields

		private readonly object _lockObject = new object();
		private readonly string[] _files;

		private int _index = -1;

		#endregion

		#region Public constructor

		public DirectoryEnumerator(string directory)
		{
			if (directory == null || !Directory.Exists(directory))
				throw new ArgumentException("Directory doesn't exist");

			Dir = directory;
			_files = Directory.GetFiles(Dir);
		}

		#endregion

		#region Public properties

		public string Dir { get; private set; }

		#endregion

		#region ITemplateLoader members

		public void BeginLoad()
		{
			lock (_lockObject)
			{
				if (_index != -1) throw new InvalidOperationException();
				_index = 0;
			}
		}

		public void EndLoad()
		{
			lock (_lockObject)
			{
				_index = -1;
			}
		}

		public bool LoadNext(out NSubject[] subjects, int n)
		{
			lock (_lockObject)
			{
				if (_index == -1) throw new InvalidOperationException();
				subjects = null;
				if (_files.Length == 0 || _files.Length <= _index) return false;

				int count = _files.Length - _index;
				count = count > n ? n : count;
				subjects = new NSubject[count];
				for (int i = 0; i < count; i++)
				{
					string file = _files[_index++];
					subjects[i] = new NSubject();
					subjects[i].SetTemplateBuffer(new NBuffer(File.ReadAllBytes(file)));
					subjects[i].Id = Path.GetFileNameWithoutExtension(file);
				}
				return true;
			}
		}

		public void Dispose()
		{
		}

		public int TemplateCount
		{
			get { return _files.Length; }
		}

		#endregion

		#region Public methods

		public override string ToString()
		{
			return Dir;
		}

		#endregion
	}
}
